cpp/src/arrow/compute/kernels/vector_cumulative_sum.cc

```diff
@@ -51,8 +51,8 @@ struct CumulativeSum {
   }
 
   Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
-    const ScalarType& start_scalar = checked_cast<const ScalarType&>(*batch[1].scalar());
-    CType start = start_scalar.value;
+    const auto& options = OptionsWrapper<CumulativeSumOptions<CType>>::Get(ctx);
+    CType start = checked_cast<const ScalarType&>(options.start).value;
```

Comment:
Consider using UnboxScalar instead: https://github.com/apache/arrow/blob/e8a1e1328f3c1bc6bf5ff7ba86057af3e37b634d/cpp/src/arrow/compute/kernels/codegen_internal.h#L325-L328

Though: we should handle when the scalar is null (`!scalar->is_valid`), when the scalar is `nullptr`, and when the scalar is not of the expected type.